<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Mgtaxa: MGT::Proj::HctApp::HctApp Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.2 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><b>MGT</b>      </li>
      <li><b>Proj</b>      </li>
      <li><b>HctApp</b>      </li>
      <li><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html">HctApp</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a>  </div>
  <div class="headertitle">
<h1>MGT::Proj::HctApp::HctApp Class Reference</h1>  </div>
</div>
<div class="contents">
<!-- doxytag: class="MGT::Proj::HctApp::HctApp" --><!-- doxytag: inherits="MGT::App::App" --><!-- startSectionHeader --><div class="dynheader">
Inheritance diagram for MGT::Proj::HctApp::HctApp:<!-- endSectionHeader --></div>
<!-- startSectionSummary --><!-- endSectionSummary --><!-- startSectionContent --><div class="dyncontent">
 <div class="center">
  <img src="classMGT_1_1Proj_1_1HctApp_1_1HctApp.png" usemap="#MGT::Proj::HctApp::HctApp_map" alt=""/>
  <map id="MGT::Proj::HctApp::HctApp_map" name="MGT::Proj::HctApp::HctApp_map">
<area href="classMGT_1_1App_1_1App.html" alt="MGT::App::App" shape="rect" coords="0,0,164,24"/>
</map>
 </div><!-- endSectionContent --></div>

<p><a href="classMGT_1_1Proj_1_1HctApp_1_1HctApp-members.html">List of all members.</a></p>
<table class="memberdecls">
<tr><td colspan="2"><h2><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#af287c10464f0a72e435a9143539e9b57">parseCmdLinePost</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#a4b8055d0c4b07f3471fe6144b978715d">getDbSql</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#a5ea6941104a171ac2101b59071212886">delDbSql</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#a46cf0c1ac64f544e4d67410038d7973a">doWork</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#ae3e76d316b2f7038a1cd287ac09417bb">pullSeq</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#a08deabbbbe9cf6ce409d3c12c0233eda">loadGenomicIdsOrgType</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#afc15e96d481f812c4c86c214e9906ee4">pullGenBankProts</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#a92bb4bc7940edd5e191c3c5b11551b1d">pullOutGroupProts</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#ac3528d040e8bd1a1973c8e792390dd8d">pullGenBankProtsCatOld</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#ade2a94c42a67d73a6cbb2749d1c45f98">loadProtIdsOrgType</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#a376527e85616bf08cebfea741171a31d">sqlPostPullSeq</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#afb6010c08ec68ec99648ce214127b200">makeSeqGenNames</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#a43ac7a7417b385cbdbee4a43571668c1">selectChromosomes</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#af5d294d63a65eb4adabe265836319f2b">selectAnyGenElement</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#aa3d5b2c37ff40ff4e20a395ca9d21a39">runCvTree</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#a67d662811454afbe66c4bd5edcf9e453">runCvTreeMat</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#a5560ee98367d57b6f7523089b1b1a202">subSampleProtIds</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#a8ab6b25c5e80fc70aa3ed8da3b61d100">loadSelProtIdsMem</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#a6d953e1abc156498a758c44c23e38271">makeCvTreeSeqInput</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#aa1182fde05522a8baea884fda36b5910">loadCvTreeNamesFile</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#a08517cd8c3de5b94c2a889b2fdf9a20d">cvMatToPhylipMat</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#a00264aff67921da9ad2121c7fc650890">loadAccToNameMem</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1HctApp_1_1HctApp.html#a15ccb60367ccadb0bd89b5ccb2dd0fe4">exportTreeDynAnnot</a></td></tr>
</table>
<hr/><a name="_details"></a><h2>Detailed Description</h2>
<div class="fragment"><pre class="fragment">App-derived class for Horizontal chromosome transfer test</pre></div> <hr/><h2>Member Function Documentation</h2>
<a class="anchor" id="a08517cd8c3de5b94c2a889b2fdf9a20d"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::cvMatToPhylipMat" ref="a08517cd8c3de5b94c2a889b2fdf9a20d" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::cvMatToPhylipMat </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Rename sequence names in the distance matrix into Phylip compatible names</pre></div> 
</div>
</div>
<a class="anchor" id="a5ea6941104a171ac2101b59071212886"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::delDbSql" ref="a5ea6941104a171ac2101b59071212886" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::delDbSql </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Call this to free a connection to SQL server if it will not be needed for extended period of time</pre></div> 
</div>
</div>
<a class="anchor" id="a46cf0c1ac64f544e4d67410038d7973a"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::doWork" ref="a46cf0c1ac64f544e4d67410038d7973a" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::doWork </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>kw</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Do the actual work.
Must be redefined in the derived classes.
Should not be called directly by the user except from doWork() in a derived class.
Should work with empty keyword dict, using only self.opt.
If doing batch submision of other App instances, must return a list of sink (final) BatchJob objects.</pre></div> 
<p>Reimplemented from <a class="el" href="classMGT_1_1App_1_1App.html#a7dbfd4f4b688baa8b586ac92f0a1b368">MGT::App::App</a>.</p>

</div>
</div>
<a class="anchor" id="a15ccb60367ccadb0bd89b5ccb2dd0fe4"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::exportTreeDynAnnot" ref="a15ccb60367ccadb0bd89b5ccb2dd0fe4" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::exportTreeDynAnnot </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Create a file in TreeDyn annotation format.
Example of such file format (taken from TreeDyn web site):
BUD2  Subcellular_loc { Bud_neck Cytoskeletal } Cellular_Role { Cell_polarity } FuncCat { GTPase_activating_protein }
BUD3  Subcellular_loc { Bud_neck } Cellular_Role { Cell_polarity } FuncCat { Unknown }
</pre></div> 
</div>
</div>
<a class="anchor" id="a4b8055d0c4b07f3471fe6144b978715d"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::getDbSql" ref="a4b8055d0c4b07f3471fe6144b978715d" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::getDbSql </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Allocate (if necessary) and return a connection to SQL server</pre></div> 
</div>
</div>
<a class="anchor" id="a00264aff67921da9ad2121c7fc650890"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::loadAccToNameMem" ref="a00264aff67921da9ad2121c7fc650890" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::loadAccToNameMem </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Load dict(acc-&gt;name) from SQL table</pre></div> 
</div>
</div>
<a class="anchor" id="aa1182fde05522a8baea884fda36b5910"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::loadCvTreeNamesFile" ref="aa1182fde05522a8baea884fda36b5910" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::loadCvTreeNamesFile </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Load names from a name file</pre></div> 
</div>
</div>
<a class="anchor" id="a08deabbbbe9cf6ce409d3c12c0233eda"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::loadGenomicIdsOrgType" ref="a08deabbbbe9cf6ce409d3c12c0233eda" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::loadGenomicIdsOrgType </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>db</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>idGenSeq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>inserterSeq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>orgType</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Load Fasta deflines for genomic sequences, index them by accession, and drop non-NC_ and all plasmids.
@todo It might be more robust to parse the GenBank file, e.g.
FEATURES             Location/Qualifiers
source          1..208369
    /organism="Bacillus cereus ATCC 10987"
    /mol_type="genomic DNA"
    /strain="ATCC 10987"
    /db_xref="ATCC:10987"
    /db_xref="taxon:222523"
    /plasmid="pBc10987"
                                                                                                           gene            join(207497..208369,1..687)

We would have to fix the gap(unk100) bug first, and also check how the "extrachromosomal" is labeled
in GB file.
</pre></div> 
</div>
</div>
<a class="anchor" id="ade2a94c42a67d73a6cbb2749d1c45f98"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::loadProtIdsOrgType" ref="ade2a94c42a67d73a6cbb2749d1c45f98" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::loadProtIdsOrgType </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>db</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>idGenSeq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>inserterSeq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>orgType</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Load Fasta deflines for protein sequences generated by pullSeq, parse and insert them into SQL table.
</pre></div> 
</div>
</div>
<a class="anchor" id="a8ab6b25c5e80fc70aa3ed8da3b61d100"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::loadSelProtIdsMem" ref="a8ab6b25c5e80fc70aa3ed8da3b61d100" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::loadSelProtIdsMem </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Return as Numpy array table created by subSampleProtIds()</pre></div> 
</div>
</div>
<a class="anchor" id="a6d953e1abc156498a758c44c23e38271"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::makeCvTreeSeqInput" ref="a6d953e1abc156498a758c44c23e38271" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::makeCvTreeSeqInput </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>accProts</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Create input FASTA files for CVTree by selecting only sequences with Acc from accProts.
Warning: this will erase existing content of the target directory.</pre></div> 
</div>
</div>
<a class="anchor" id="afb6010c08ec68ec99648ce214127b200"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::makeSeqGenNames" ref="afb6010c08ec68ec99648ce214127b200" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::makeSeqGenNames </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">For each genetic element, we generate an artificial mnemonic name.
The purpose of generating a new name is two-fold:
1. Phylip cannot handle names longer than 10 symbols
2. For each organism with multiple genetic elements, our name will reflect the
nature of that element (chromosome, plasmid,...), as well as number elements of
the same nature in the order of reducing size.
Examples:
If the organism is assigned id 123 and it has only one chromosome,
that chromosome will be called t0123.
If that organism has two chromosomes, they will be called t0123C1, and t0123C2,
with C1 being the larger one.
The disparity between one- and multi-chromosome naming schemes is designed to make
multi-chromosomal organisms to visibly stand out.
For plasmids, names will be like t0123P1.
For all other genetic elements - like t0123?1 where ? is the first letter of 
genel field of seq_gen table.
The names are saved into an SQL table name_gen.
</pre></div> 
</div>
</div>
<a class="anchor" id="af287c10464f0a72e435a9143539e9b57"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::parseCmdLinePost" ref="af287c10464f0a72e435a9143539e9b57" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::parseCmdLinePost </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>klass</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>options</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>args</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>parser</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Optionally modify options and args in-place.
Called at the end of parseCmdLine to allow the derived classes customizing the option processing.
@param options options returned by OptionParser and converted to Struct object
@param args args returned by OptionParser
@param parser OptionParser object used to parse the command line - needed here to call its error() method
if necessary.
options should be modified in place by this method</pre></div> 
<p>Reimplemented from <a class="el" href="classMGT_1_1App_1_1App.html#a24772e34d10ff65d64f403cc54e8c423">MGT::App::App</a>.</p>

</div>
</div>
<a class="anchor" id="afc15e96d481f812c4c86c214e9906ee4"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::pullGenBankProts" ref="afc15e96d481f812c4c86c214e9906ee4" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::pullGenBankProts </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>orgType</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>idsGen</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Scan a set of GenBank protein files and output a multi-FASTA file with each protein as a separate record and chromosome encoded in defline.
The defline will look like:
&gt;YP_089573|221988_NC_006300.1
as in &gt;&lt;protein_acc&gt;|&lt;taxid&gt;_&lt;chrmosome_acc&gt;
</pre></div> 
</div>
</div>
<a class="anchor" id="ac3528d040e8bd1a1973c8e792390dd8d"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::pullGenBankProtsCatOld" ref="ac3528d040e8bd1a1973c8e792390dd8d" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::pullGenBankProtsCatOld </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Scan a set of GenBank protein files and output a multi-FASTA file with all proteins in one chromosome concatenated.
@todo We cannot insert spacers between proteins, because CvTree will not care. So we get chimeric k-mers.</pre></div> 
</div>
</div>
<a class="anchor" id="a92bb4bc7940edd5e191c3c5b11551b1d"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::pullOutGroupProts" ref="a92bb4bc7940edd5e191c3c5b11551b1d" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::pullOutGroupProts </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Convert RefSeq per-chromosome .faa file into .faa file with our defline format.
The defline will look like:
&gt;YP_089573|221988_NC_006300.1
as in &gt;&lt;protein_acc&gt;|&lt;taxid&gt;_&lt;chrmosome_acc&gt;
</pre></div> 
</div>
</div>
<a class="anchor" id="ae3e76d316b2f7038a1cd287ac09417bb"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::pullSeq" ref="ae3e76d316b2f7038a1cd287ac09417bb" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::pullSeq </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Convert original RefSeq Fasta and GenBank files into SQL tables and protein Fasta files</pre></div> 
</div>
</div>
<a class="anchor" id="aa3d5b2c37ff40ff4e20a395ca9d21a39"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::runCvTree" ref="aa3d5b2c37ff40ff4e20a395ca9d21a39" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::runCvTree </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Run cvtree executable that generates the composition vectors.
Warning: this clears existing content of cvVecDir</pre></div> 
</div>
</div>
<a class="anchor" id="a67d662811454afbe66c4bd5edcf9e453"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::runCvTreeMat" ref="a67d662811454afbe66c4bd5edcf9e453" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::runCvTreeMat </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Run cvtree script that builds the distance matrix from precomputed composition vectors.
</pre></div> 
</div>
</div>
<a class="anchor" id="af5d294d63a65eb4adabe265836319f2b"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::selectAnyGenElement" ref="af5d294d63a65eb4adabe265836319f2b" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::selectAnyGenElement </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>minProtSeqLen</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Select genomic elements of any type longer than cutoff value for the distance matrix calculation</pre></div> 
</div>
</div>
<a class="anchor" id="a43ac7a7417b385cbdbee4a43571668c1"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::selectChromosomes" ref="a43ac7a7417b385cbdbee4a43571668c1" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::selectChromosomes </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>minProtSeqLen</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Select chromosomes subset for the distance matrix calculation</pre></div> 
</div>
</div>
<a class="anchor" id="a376527e85616bf08cebfea741171a31d"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::sqlPostPullSeq" ref="a376527e85616bf08cebfea741171a31d" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::sqlPostPullSeq </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Create various derived tables after pulling sequences</pre></div> 
</div>
</div>
<a class="anchor" id="a5560ee98367d57b6f7523089b1b1a202"></a><!-- doxytag: member="MGT::Proj::HctApp::HctApp::subSampleProtIds" ref="a5560ee98367d57b6f7523089b1b1a202" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::HctApp::HctApp::subSampleProtIds </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>maxProtSeqLen</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Select a subset of proteins from each genomic sequence constrained by a sum of protein lengths.
The use case: we want to have each chromosome represented by an equal length sample, to make
sure that the phylogenetic tree we build is not influenced by length differences.</pre></div> 
</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>mgtaxa/MGT/Proj/HctApp.py</li>
</ul>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Thu Jan 20 2011 17:14:45 for Mgtaxa by&#160;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.2 </small></address>
</body>
</html>
