<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=iso-8859-1">
<title>Mgtaxa: MGT::CollectTaxa::TaxaCollector Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.4.6 -->
<div class="tabs">
  <ul>
    <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
    <li id="current"><a href="annotated.html"><span>Classes</span></a></li>
    <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
  </ul></div>
<div class="tabs">
  <ul>
    <li><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
    <li><a href="hierarchy.html"><span>Class&nbsp;Hierarchy</span></a></li>
    <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
  </ul></div>
<div class="nav">
<b>MGT</b>::<b>CollectTaxa</b>::<a class="el" href="classMGT_1_1CollectTaxa_1_1TaxaCollector.html">TaxaCollector</a></div>
<h1>MGT::CollectTaxa::TaxaCollector Class Reference</h1><!-- doxytag: class="MGT::CollectTaxa::TaxaCollector" --><a href="classMGT_1_1CollectTaxa_1_1TaxaCollector-members.html">List of all members.</a><table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1CollectTaxa_1_1TaxaCollector.html#555057a3450389bb18febdd402bbb519">loadTaxTables</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1CollectTaxa_1_1TaxaCollector.html#3b7a2ff61d5f6daa0683fb111f3eea09">loadSeqToHdf</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1CollectTaxa_1_1TaxaCollector.html#2ee55902de16cdb0c273c548c7e3441b">indexHdfSeq</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1CollectTaxa_1_1TaxaCollector.html#8a28f98f995a6eca548c9dd98c76a662">indexHdfActiveSeq</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1CollectTaxa_1_1TaxaCollector.html#a09423f942b8ac1961b87ec74ed634f6">reportStat</a></td></tr>

<tr><td colspan="2"><br><h2>Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1CollectTaxa_1_1TaxaCollector.html#e00d138c73698946f22cafa52fb0c5b0">dbSqltblNodes</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">If we get a header record with two gi's like: ## &gt;gi|23455713|ref|NC_004301.1| Enterobacteria phage FI, complete genome &gt;gi|15183|emb|X07489.1| Bacteriophage SP genomic RNA ## then we insert the second gi along with db ("emb" here) into the seq_multi_id table, related to the record in 'seq' by ## "seq_multi_id.id_seq = seq.id".  <a href="#e00d138c73698946f22cafa52fb0c5b0"></a><br></td></tr>
</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
<div class="fragment"><pre class="fragment">Collects data about taxonomically known sequence for training the classifier.
Process the NCBI BLAST DB files by calling fastacmd.
Aggregate, prioritize and partially remove redundancy along the taxonomy ids.</pre></div> 
<p>
<hr><h2>Member Function Documentation</h2>
<a class="anchor" name="8a28f98f995a6eca548c9dd98c76a662"></a><!-- doxytag: member="MGT::CollectTaxa::TaxaCollector::indexHdfActiveSeq" ref="8a28f98f995a6eca548c9dd98c76a662" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::CollectTaxa::TaxaCollector::indexHdfActiveSeq           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Create HDF index dataset for active sequence (from act_seq table).</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="2ee55902de16cdb0c273c548c7e3441b"></a><!-- doxytag: member="MGT::CollectTaxa::TaxaCollector::indexHdfSeq" ref="2ee55902de16cdb0c273c548c7e3441b" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::CollectTaxa::TaxaCollector::indexHdfSeq           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Create SQL tables that index sequence in HDF dataset.</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="3b7a2ff61d5f6daa0683fb111f3eea09"></a><!-- doxytag: member="MGT::CollectTaxa::TaxaCollector::loadSeqToHdf" ref="3b7a2ff61d5f6daa0683fb111f3eea09" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::CollectTaxa::TaxaCollector::loadSeqToHdf           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Load sequence data for all records in 'seq' table into HDF dataset.</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="555057a3450389bb18febdd402bbb519"></a><!-- doxytag: member="MGT::CollectTaxa::TaxaCollector::loadTaxTables" ref="555057a3450389bb18febdd402bbb519" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::CollectTaxa::TaxaCollector::loadTaxTables           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Create and fill all initial tables with taxa tree data.
It only needs NCBI taxonomy dump files.
It assigns dummy zero values to seq_len and seq_len_total attributes.
Result: original tree dumf files as well as our tree object with nested set index etc 
is saved in SQL DB</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="a09423f942b8ac1961b87ec74ed634f6"></a><!-- doxytag: member="MGT::CollectTaxa::TaxaCollector::reportStat" ref="a09423f942b8ac1961b87ec74ed634f6" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::CollectTaxa::TaxaCollector::reportStat           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Output a report that shows a high-level overview of data.</pre></div>     </td>
  </tr>
</table>
<hr><h2>Member Data Documentation</h2>
<a class="anchor" name="e00d138c73698946f22cafa52fb0c5b0"></a><!-- doxytag: member="MGT::CollectTaxa::TaxaCollector::dbSqltblNodes" ref="e00d138c73698946f22cafa52fb0c5b0" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"><a class="el" href="classMGT_1_1CollectTaxa_1_1TaxaCollector.html#e00d138c73698946f22cafa52fb0c5b0">MGT::CollectTaxa::TaxaCollector::dbSqltblNodes</a>          </td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
If we get a header record with two gi's like: ## &gt;gi|23455713|ref|NC_004301.1| Enterobacteria phage FI, complete genome &gt;gi|15183|emb|X07489.1| Bacteriophage SP genomic RNA ## then we insert the second gi along with db ("emb" here) into the seq_multi_id table, related to the record in 'seq' by ## "seq_multi_id.id_seq = seq.id". 
<p>
self.dbSql.ddl(""" create table seq_multi_id ( id_seq integer, gi bigint, acc_db char(3) ) """, dropList=["table seq_multi_id"]) def loadSeqNCBI(self,db,inserterSeq,inserterSeqMultiId,inserterSeqHdr): print "Processing BLAST DB " + db.db inp = self.blastDb.fastaReader(dbName=db.db,defLineTargetOnly=False) iRec = 0 for rec in inp.records(): idSeq = self.idGenSeq() title = rec.header()[1:-1] remove '&gt;' and '<br>
' (gifld,gi,acc_db,acc,txt) = title.split('|',4) assert gifld == 'gi' gi = int(gi) defLine2 = txt.split("&gt;gi|") if len(defLine2) &gt; 1: txt = defLine2[0] (gi2,acc_db2,dummy) = defLine2[1].split('|',2) gi2 = int(gi2) inserterSeqMultiId((idSeq,gi2,acc_db2)) try: taxid = int(self.gi2taxa[gi]) except KeyError: print "Warning: Taxid not found for gi: "+gi taxid = 0 # We are not interested in taxonomically unassigned sequence. # If we late change this, we also need to modify # exportIdsForSeqDb() where we currently do the inner join # with tax nodes table if taxid != 0: Accession number formats are described here: http://www.ncbi.nlm.nih.gov/Sequin/acc.html WGS id might be 'AAAA00000000' or 'NZ_AAAA00000000', both cases seen in wgs db kind = '' acc_sfx = acc if acc_sfx[2] == '_': kind = acc_sfx[:2] acc_sfx = acc_sfx[3:] project = '' if acc_sfx[:4].isalpha() and acc_sfx[5].isdigit(): project = acc_sfx[:4] seqLen = rec.seqLen() values = (idSeq,gi,taxid,db.id,project,seqLen,acc_db,acc,kind) values = [ str(x) for x in values ] inserterSeq(values) values = (idSeq,gi,title[:self.fastaHdrSqlLen]) inserterSeqHdr(values) if iRec % 50000 == 0: print db.db, title, taxid, iRec, seqLen if iRec &gt;= 500000: # break iRec += 1 inp.close() def tmp_loadSeqHdr(self): idGenSeq = IntIdGenerator() db = self.dbSql db.ddl(""" create table seq_hdr ( id_seq integer, gi bigint, hdr varchar(i) ) """ % (self.fastaHdrSqlLen,), dropList=["table seq_hdr"]) reCutLen = re.compile(r"(.*):[0-9]+") inserterSeqHdr = db.makeBulkInserterFile(table='seq_hdr',bufLen=500000,workDir=self.tmpDir) inp = open("all.hdr",'r') for rec in inp: idSeq = idGenSeq() title = reCutLen.search(rec[1:-1]).group(1) remove '&gt;' and '<br>
', then cut the " len:0000000" suffix (gifld,gi,txt) = title.split('|',2) assert gifld == 'gi' gi = int(gi) values = (idSeq,gi,title) inserterSeqHdr(values) inserterSeqHdr.flush() db.createIndices(table="seq_hdr", names=["gi"], primary="id_seq") db.ddl("analyze table seq_hdr",ifDialect="mysql") def createFullTextIndexSeqHeader(self): """Build a full text index for sequence FASTA hdeaders in 'seq_hdr' table (currently implemented only for MySQL back-end). Warning: it took 1 hr for 28M records.""" db = self.dbSql db.ddl("alter table seq_hdr add fulltext index hdr(hdr)",ifDialect="mysql",dropList=["index hdr on seq_hdr"]) def loadRefseqAcc(self): self.dbSql.ddl("""\     </td>
  </tr>
</table>
<hr>The documentation for this class was generated from the following file:<ul>
<li>mgtaxa/MGT/CollectTaxa.py</ul>
<hr size="1"><address style="align: right;"><small>Generated on Sat May 8 15:32:20 2010 for Mgtaxa by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.4.6 </small></address>
</body>
</html>
