<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=iso-8859-1">
<title>Mgtaxa: MGT::TreeSamplerApp::TreeSamplerApp Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.4.6 -->
<div class="tabs">
  <ul>
    <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
    <li id="current"><a href="annotated.html"><span>Classes</span></a></li>
    <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
  </ul></div>
<div class="tabs">
  <ul>
    <li><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
    <li><a href="hierarchy.html"><span>Class&nbsp;Hierarchy</span></a></li>
    <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
  </ul></div>
<div class="nav">
<b>MGT</b>::<b>TreeSamplerApp</b>::<a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html">TreeSamplerApp</a></div>
<h1>MGT::TreeSamplerApp::TreeSamplerApp Class Reference</h1><!-- doxytag: class="MGT::TreeSamplerApp::TreeSamplerApp" --><a href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp-members.html">List of all members.</a><table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#c588e0fcdc5a9d1b22154a8f35c782c4">getTableName</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#d4f8262b0ce121334095cc4b7f46a0cd">getFilePath</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#8f75875ddaf1064a2489c44372e4f83c">mkDbSampleCounts</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#6a9e77ab6b21d5715f55116e24fd8b10">loadSampleCountsMem</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#e556261d17aaea7f85f3ece1e5fead31">setIsUnderUnclassMem</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#52b3f765fe853bfbef168f58fd6e82b3">markTrainable</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#c136299579252c2963a8538cab247491">unused_mkDbTestingTaxa</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#16a05eafe8d7d7198f5d257790d38f49">unused_lowRankTestPairs</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#e38d8d606cf2d1d8d350cf621150edde">unused_sampleTestPairs</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#96894d8a1c0ba001c7eef7138c09a090">isTestTerm</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#7667e4d3ace81327f21e38677561fa89">selectTestTaxa</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#d2b68e3bae905e7a57aaea4b074b2c3a">markTraining</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#68f818e4663b8617d9883e3166909479">checkPostMarkTraining</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#06f92038152e25101fe929b8e52368df">statTestTaxa</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#0356aa918e89b278d4597d691397bcf9">writeTraining</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#3f52ffe9992bd8cdfbf638af3bb7eb00">pickRandomSampleCounts</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#0c1ad7fe84e6dee1345d78db4e13a4f9">writeTesting</a></td></tr>

<tr><td colspan="2"><br><h2>Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="ec4359af608cb0b048cefca780d3c06a"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::idLabRecs" ref="ec4359af608cb0b048cefca780d3c06a" args="" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#ec4359af608cb0b048cefca780d3c06a">idLabRecs</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Try to allocate enough idLab records for both training and testing, but wrap it in ArrayAppender just in case we reach the end at some point. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="7cd5ff20b7fc98ae10c0e021f5cb048f"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::labToName" ref="7cd5ff20b7fc98ae10c0e021f5cb048f" args="" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#7cd5ff20b7fc98ae10c0e021f5cb048f">labToName</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">labToName items that are always present, others will be added later - "rj" reject group, always 0; "bg" background samples, always 1 <br></td></tr>
<tr><td colspan="2"><br><h2>Static Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="9a09f78fce7160bc74f360d3cfa60b36"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::termTopRankSeq" ref="9a09f78fce7160bc74f360d3cfa60b36" args="" -->
string&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#9a09f78fce7160bc74f360d3cfa60b36">termTopRankSeq</a> = &quot;_sequence&quot;</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Special constant to control splitting - will cause splitting along nodes with directly attached sequence. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="bfa5e13cf41d156501faa942172d3609"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::termTopRankLeaf" ref="bfa5e13cf41d156501faa942172d3609" args="" -->
string&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1TreeSamplerApp_1_1TreeSamplerApp.html#bfa5e13cf41d156501faa942172d3609">termTopRankLeaf</a> = &quot;_leaf&quot;</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Special constant to control splitting - will cause splitting along leaf nodes. <br></td></tr>
</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
<div class="fragment"><pre class="fragment">Sample chunks of available training sequence by concatenating all sequences for each taxid.
Concatenating (as opposed to chunking each sequence individually) greatly simplifies random
sampling of taxid chunks w/o replacement and increases the amount of sample data for small
sequences. It creates chimeric k-mer vectors however (but not chimeric k-mers because we insert
spacers).</pre></div> 
<p>
<hr><h2>Member Function Documentation</h2>
<a class="anchor" name="68f818e4663b8617d9883e3166909479"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::checkPostMarkTraining" ref="68f818e4663b8617d9883e3166909479" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::checkPostMarkTraining           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Consistency check after a call to markTraining().</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="d4f8262b0ce121334095cc4b7f46a0cd"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::getFilePath" ref="d4f8262b0ce121334095cc4b7f46a0cd" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::getFilePath           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>name</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Same as getTableName() but for files.</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="c588e0fcdc5a9d1b22154a8f35c782c4"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::getTableName" ref="c588e0fcdc5a9d1b22154a8f35c782c4" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::getTableName           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>name</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Given a stem table name, return full table name unique for this object (prefix+name).
@param name - stem name
</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="96894d8a1c0ba001c7eef7138c09a090"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::isTestTerm" ref="96894d8a1c0ba001c7eef7138c09a090" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::isTestTerm           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>node</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Return True if the node is designated as testing terminal.
If True, the node and its subtree can be used only in its entirety either for training or for testing split.
Currently the implementation of this methods checks if node's rank is present in self.opt.termTopRanks,
with the two special names allowed in self.opt.termTopRanks: 
1) when termTopRanks list contains element equal to self.termTopRankSeq constant, we return True for
every node that has directly attached sequence.
2) when termTopRanks list contains element equal to self.termTopRankLeaf constant, we return True for
every leaf node (without children).
The typical use to have either actual ranks in self.opt.termTopRanks, or either one of the special
constants.
Redefine this method in derived classes to use some more complicated splitting criterion.
For the default implementation, if for example, we return True for species nodes, then sequences under 
species will be never split between training and testing. That ensures proper validation of a claim 
that testing accuracy is provided for the dicovery of previously unseen species.
@note PhyloPythia 2006 paper used splits as the level of strains (or whatever the leaf sequence
taxonomy node was for).
@attention It is often the case that a species node will have directly attached sequence, while
also having strains under it with their own sequence. Most of the time, the species level sequence
is not genome-wide sequencing output, but rather sequences for some specific genes. There examples
when it is a full plasmid sequence, e.g. Aeromonas salmonicida taxid 645.
If we ignore in sample database construction everything but WGS and full genomes, and drop plasmids,
we are probably safe.
However, dealing with it in a general case probably makes sense if just pick leaf nodes as testing
terminal and ignore any sequence attached to their upper nodes.</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="6a9e77ab6b21d5715f55116e24fd8b10"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::loadSampleCountsMem" ref="6a9e77ab6b21d5715f55116e24fd8b10" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::loadSampleCountsMem           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Transfer sample counts from DB into in-memory taxonomy tree object.</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="52b3f765fe853bfbef168f58fd6e82b3"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::markTrainable" ref="52b3f765fe853bfbef168f58fd6e82b3" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::markTrainable           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Mark all nodes that have enough samples for training with and without validation.
When we train for subsequent validation with testing set, we cannot use any 'unclassified'
samples because they might actually represent the excluded testing set. On the other
hand, there is no reason not to use any 'unclassified' samples under a 'classified'
node for the final re-training.
@pre loadSampleCountsMem() and setInUnderUnclassMem() were called
@post in-memory taxaTree nodes have the following attributes set:
    isTrainable - trainable with 'unclassifed' samples included
    isTrainableTest - trainable for validation. A node can still be excluded from
    testing later if extracting testing set will leave too few samples for training,
    but it will be trained during testing stage - just will not have any true positive 
    testing samples.</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="d2b68e3bae905e7a57aaea4b074b2c3a"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::markTraining" ref="d2b68e3bae905e7a57aaea4b074b2c3a" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::markTraining           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Mark final training state for each node - 'is_class' and 'n_mclass' attributes.
Nodes that are not 'is_class' are ignored during training.
'is_class' means that a node can be used as a label in training the
multiclass classifier for the parent node.
'n_mclass' is also set to the number of classes below the node.
'samp_n_max' is the maximum leaf sample count under this node.
This method must be called after selectTestTaxa() and uses node attributes set by
that method.</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="8f75875ddaf1064a2489c44372e4f83c"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::mkDbSampleCounts" ref="8f75875ddaf1064a2489c44372e4f83c" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::mkDbSampleCounts           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Make DB tables with sample counts per taxonomy id</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="3f52ffe9992bd8cdfbf638af3bb7eb00"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::pickRandomSampleCounts" ref="3f52ffe9992bd8cdfbf638af3bb7eb00" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::pickRandomSampleCounts           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>node</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Randomly pick counts of available samples from taxonomy ids below the node.
@return dictionary { taxid -&gt; count }.
Should be called from writeTraining() which sets up sampNTrainTot and sampNTrainSelf
node property arrays.</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="7667e4d3ace81327f21e38677561fa89"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::selectTestTaxa" ref="7667e4d3ace81327f21e38677561fa89" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::selectTestTaxa           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Select nodes that will be used entirely for testing.
The method aims to achieve these goals:
- create a uniform random sampling of the tree
- sample as deep into the hierarchy as possible
- select entire genera or entire species if there is no genera above them
(we call such nodes "terminal" here)
This is done recursively: for a given node: 
a) all taxa selected as testing for its non-terminal children is also selected
b) and then terminal children are added at random until we reach the limit on 
testing samples set for this node.
The major alternative would be to do the (b) selection from a list of all
so far non-selected terminal subnodes in an entire sub-tree of a given node
(the method would have to return such a list from each recursive call, with the
parent concatenating the lists). This would provided better chances to reach the
testing ratio target for higher order taxa. However, this would also run a danger
of a non-uniform training and deteriorating perfomance. For example, if we have
a subtree (A,(B,C),D) and B was selected as testing for (B,C) subtree, we could also
select C as testing for (A,(B,C),D) subtree (it would still be used as training
for (B,C). However, we would have no trainig samples from (B,C) when trainig (A,(B,C),D),
which can become an issue.
@post Attributes are set for each node:
    - is_test - if True, entire subtree is selected as testing sample. Set for every node
    in such a subtree.
    - has_test - if True, node has enough testing samples, but it is not 'is_test' itself.
    - samp_n_tot_test - number of testing samples for this node
    - samp_n_tot_train - number of training samples for this node.
    - isTestTerm - split boundaries (entire subtrees can have either True or False)
Note that the presence of "unclassified" sequence still creates problems:
selection of even one testing sample excludes from training all unclassified nodes
which are immediate children of all nodes in testing sample lineage. Thus, any of these nodes
can become non-trainable even if it is trainable w/o testing. However, we need to accumulate
the testing sequence for use by upper nodes, so this seems unavoidable. To check the effect
of this tradeoff, temporarily make TaxaNode.isUnclassified() to always return False and re-run
the testing/training selection. So far this experiment shown little impact.</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="e556261d17aaea7f85f3ece1e5fead31"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::setIsUnderUnclassMem" ref="e556261d17aaea7f85f3ece1e5fead31" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::setIsUnderUnclassMem           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Set an attribute 'isUnderUnclass' for in-memory tree nodes.
It flags all nodes that have "unclassified" super-node somewhere in their lineage.</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="06f92038152e25101fe929b8e52368df"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::statTestTaxa" ref="06f92038152e25101fe929b8e52368df" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::statTestTaxa           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Save SQL statistics about testing/training set selection.
Call after markTraining()</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="16a05eafe8d7d7198f5d257790d38f49"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::unused_lowRankTestPairs" ref="16a05eafe8d7d7198f5d257790d38f49" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::unused_lowRankTestPairs           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>rank</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>superRank</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Return set of unique pairs (testing node,first supernode).
@param rank - entire nodes of this rank will be selected as testing samples (e.g. genus).
@param superRank - a second member of each pair will have at least this rank (e.g. family).
'superRank' nodes might not be 'isTrainable'. That should be filtered later.</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="c136299579252c2963a8538cab247491"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::unused_mkDbTestingTaxa" ref="c136299579252c2963a8538cab247491" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::unused_mkDbTestingTaxa           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>rank</em> = <code>&quot;genus&quot;</code>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>superRank</em> = <code>&quot;family&quot;</code></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Select complete 'rank' nodes uniformly distributed across 'superRank' super-nodes.</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="e38d8d606cf2d1d8d350cf621150edde"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::unused_sampleTestPairs" ref="e38d8d606cf2d1d8d350cf621150edde" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::unused_sampleTestPairs           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>pairs</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Randomly sample a list of (testing node,first supernode) pairs obtained with lowRankTestPairs().</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="0c1ad7fe84e6dee1345d78db4e13a4f9"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::writeTesting" ref="0c1ad7fe84e6dee1345d78db4e13a4f9" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::writeTesting           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>svmWriter</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Write testing data set.
@pre Nodes have the following attributes: 
is_class - True for nodes that can serve as classes in SVM
n_mclass - number of trainable sub-classes
samp_n_tot_test - number of testing samples in a subtree
is_test - True if a subtree is used for testing
This method must be called after markTraining() sets needed node attributes.
</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="0356aa918e89b278d4597d691397bcf9"></a><!-- doxytag: member="MGT::TreeSamplerApp::TreeSamplerApp::writeTraining" ref="0356aa918e89b278d4597d691397bcf9" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::TreeSamplerApp::TreeSamplerApp::writeTraining           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>trNodes</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>svmWriter</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Write training data set for each eligible node.
@pre Nodes have the following attributes: 
is_class - True for nodes that can serve as classes in SVM
n_mclass - number of trainable sub-classes
samp_n_tot_test - number of testing samples in a subtree
samp_n_tot_train - number of training samples in a subtree (excluding testing)
is_test - True if a subtree is used for testing
@post Data sets ready for feature generation are saved.
@post taxaTree.getRootNode().setIndex() resets internal node index.
This method must be called after markTraining() sets needed node attributes.
</pre></div>     </td>
  </tr>
</table>
<hr>The documentation for this class was generated from the following file:<ul>
<li>mgtaxa/MGT/TreeSamplerApp.py</ul>
<hr size="1"><address style="align: right;"><small>Generated on Sat May 8 15:32:21 2010 for Mgtaxa by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.4.6 </small></address>
</body>
</html>
