<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Mgtaxa: MGT::Proj::CrisprMooreApp::CrisprMooreApp Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<!-- Generated by Doxygen 1.7.2 -->
<div class="navigation" id="top">
  <div class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
    </ul>
  </div>
  <div class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
  <div class="navpath">
    <ul>
      <li><b>MGT</b>      </li>
      <li><b>Proj</b>      </li>
      <li><b>CrisprMooreApp</b>      </li>
      <li><a class="el" href="classMGT_1_1Proj_1_1CrisprMooreApp_1_1CrisprMooreApp.html">CrisprMooreApp</a>      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a>  </div>
  <div class="headertitle">
<h1>MGT::Proj::CrisprMooreApp::CrisprMooreApp Class Reference</h1>  </div>
</div>
<div class="contents">
<!-- doxytag: class="MGT::Proj::CrisprMooreApp::CrisprMooreApp" -->
<p>### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##  
<a href="#_details">More...</a></p>

<p><a href="classMGT_1_1Proj_1_1CrisprMooreApp_1_1CrisprMooreApp-members.html">List of all members.</a></p>
<table class="memberdecls">
<tr><td colspan="2"><h2><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprMooreApp_1_1CrisprMooreApp.html#ad9cea2d50343b80db7f0676196980c13">loadMic</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprMooreApp_1_1CrisprMooreApp.html#ad6885affd3891a74569d4ab10ae13e99">loadMooreSql</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprMooreApp_1_1CrisprMooreApp.html#addfc1d2c388e7a029ac1c09645ba39b3">loadClosedSql</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprMooreApp_1_1CrisprMooreApp.html#a0a33bb195aa4bf491ef0253016584fff">mergeMicPathsSql</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprMooreApp_1_1CrisprMooreApp.html#a8a88067c153fad22639de4a2a4439354">makeMicNucFasta</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprMooreApp_1_1CrisprMooreApp.html#a8377aaf3bdaef17659c026a6ad1b2cc7">loadMicProtHitsSql</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprMooreApp_1_1CrisprMooreApp.html#a151ed587dbef11c386ea3693905df338">loadMicProtAnnotCrisprSql</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprMooreApp_1_1CrisprMooreApp.html#a4f8e2e4b6ea169ffd8223cee25fb3849">annotMicCrisprGb</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprMooreApp_1_1CrisprMooreApp.html#a1ec0011894da12b0fcc82160800edf6b">plotCrispr</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprMooreApp_1_1CrisprMooreApp.html#aefe84e1be9e86db132c1f8a4f73df957">plotCrisprOneBioGff</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprMooreApp_1_1CrisprMooreApp.html#afeb2d95031e143a6751434cf9717a74c">plotCrisprOneGff</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprMooreApp_1_1CrisprMooreApp.html#a8642fa781cc7d3c963a68c14d3cebd52">plotCrisprOneBioGd</a></td></tr>
</table>
<hr/><a name="_details"></a><h2>Detailed Description</h2>
<p>### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ## </p>
<p>See COPYING file distributed along with the MGTAXA package for the copyright and license terms.</p>
<p>### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ## </p>
<div class="fragment"><pre class="fragment">App-derived class for CRISPR array processing in "Moore genomes" dataset</pre></div> <hr/><h2>Member Function Documentation</h2>
<a class="anchor" id="a4f8e2e4b6ea169ffd8223cee25fb3849"></a><!-- doxytag: member="MGT::Proj::CrisprMooreApp::CrisprMooreApp::annotMicCrisprGb" ref="a4f8e2e4b6ea169ffd8223cee25fb3849" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::CrisprMooreApp::CrisprMooreApp::annotMicCrisprGb </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Add CRISPR arrays as annotation to Genbank format files</pre></div> 
</div>
</div>
<a class="anchor" id="addfc1d2c388e7a029ac1c09645ba39b3"></a><!-- doxytag: member="MGT::Proj::CrisprMooreApp::CrisprMooreApp::loadClosedSql" ref="addfc1d2c388e7a029ac1c09645ba39b3" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::CrisprMooreApp::CrisprMooreApp::loadClosedSql </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Load meta-info about Genbank non-Moore closed marine genomes into SQL tables</pre></div> 
</div>
</div>
<a class="anchor" id="ad9cea2d50343b80db7f0676196980c13"></a><!-- doxytag: member="MGT::Proj::CrisprMooreApp::CrisprMooreApp::loadMic" ref="ad9cea2d50343b80db7f0676196980c13" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::CrisprMooreApp::CrisprMooreApp::loadMic </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Load data from Mic genomes necessary to run our CRISPR pipeline on them</pre></div> 
</div>
</div>
<a class="anchor" id="a151ed587dbef11c386ea3693905df338"></a><!-- doxytag: member="MGT::Proj::CrisprMooreApp::CrisprMooreApp::loadMicProtAnnotCrisprSql" ref="a151ed587dbef11c386ea3693905df338" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::CrisprMooreApp::CrisprMooreApp::loadMicProtAnnotCrisprSql </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Create SQL table with the most definitive name for CAS genes and their coordinates</pre></div> 
</div>
</div>
<a class="anchor" id="a8377aaf3bdaef17659c026a6ad1b2cc7"></a><!-- doxytag: member="MGT::Proj::CrisprMooreApp::CrisprMooreApp::loadMicProtHitsSql" ref="a8377aaf3bdaef17659c026a6ad1b2cc7" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::CrisprMooreApp::CrisprMooreApp::loadMicProtHitsSql </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Load reliable protein HMM hits for Moore and other closed genomes into SQL tables</pre></div> 
</div>
</div>
<a class="anchor" id="ad6885affd3891a74569d4ab10ae13e99"></a><!-- doxytag: member="MGT::Proj::CrisprMooreApp::CrisprMooreApp::loadMooreSql" ref="ad6885affd3891a74569d4ab10ae13e99" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::CrisprMooreApp::CrisprMooreApp::loadMooreSql </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Load meta-info about Moore genomes into SQL tables</pre></div> 
</div>
</div>
<a class="anchor" id="a8a88067c153fad22639de4a2a4439354"></a><!-- doxytag: member="MGT::Proj::CrisprMooreApp::CrisprMooreApp::makeMicNucFasta" ref="a8a88067c153fad22639de4a2a4439354" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::CrisprMooreApp::CrisprMooreApp::makeMicNucFasta </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Export microbial GB nucl sequence into FASTA, store seq IDs into SQL table along the way, and split files for Piler run</pre></div> 
</div>
</div>
<a class="anchor" id="a0a33bb195aa4bf491ef0253016584fff"></a><!-- doxytag: member="MGT::Proj::CrisprMooreApp::CrisprMooreApp::mergeMicPathsSql" ref="a0a33bb195aa4bf491ef0253016584fff" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::CrisprMooreApp::CrisprMooreApp::mergeMicPathsSql </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Merge tables *_path with files paths for different sources of data, such as Moore and other Genbank marine genomes.
@post Results are in mic_path table</pre></div> 
</div>
</div>
<a class="anchor" id="a1ec0011894da12b0fcc82160800edf6b"></a><!-- doxytag: member="MGT::Proj::CrisprMooreApp::CrisprMooreApp::plotCrispr" ref="a1ec0011894da12b0fcc82160800edf6b" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::CrisprMooreApp::CrisprMooreApp::plotCrispr </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>&#160;)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Use Bio.Graphics.GenomeDiagram to generate pictures of CRISPR genes and arrays from pre-processed Genbank files</pre></div> 
</div>
</div>
<a class="anchor" id="a8642fa781cc7d3c963a68c14d3cebd52"></a><!-- doxytag: member="MGT::Proj::CrisprMooreApp::CrisprMooreApp::plotCrisprOneBioGd" ref="a8642fa781cc7d3c963a68c14d3cebd52" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::CrisprMooreApp::CrisprMooreApp::plotCrisprOneBioGd </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>seqRec</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>outGraphFileRoot</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Use Bio.Graphics.GenomeDiagram to generate pictures of CRISPR genes and arrays from one SeqRecord from a pre-processed Genbank file</pre></div> 
</div>
</div>
<a class="anchor" id="aefe84e1be9e86db132c1f8a4f73df957"></a><!-- doxytag: member="MGT::Proj::CrisprMooreApp::CrisprMooreApp::plotCrisprOneBioGff" ref="aefe84e1be9e86db132c1f8a4f73df957" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::CrisprMooreApp::CrisprMooreApp::plotCrisprOneBioGff </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>seqRec</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>outGraphFileRoot</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Generate GFF files of CRISPR genes and arrays from one SeqRecord from a pre-processed Genbank file, using Bio sandbox module</pre></div> 
</div>
</div>
<a class="anchor" id="afeb2d95031e143a6751434cf9717a74c"></a><!-- doxytag: member="MGT::Proj::CrisprMooreApp::CrisprMooreApp::plotCrisprOneGff" ref="afeb2d95031e143a6751434cf9717a74c" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def MGT::Proj::CrisprMooreApp::CrisprMooreApp::plotCrisprOneGff </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>seqRec</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"> <em>outGraphFileRoot</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<div class="fragment"><pre class="fragment">Generate GFF files and graphics for CRISPR genes and arrays from one SeqRecord from a pre-processed Genbank file</pre></div> 
</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>mgtaxa/MGT/Proj/CrisprMooreApp.py</li>
</ul>
</div>
<hr class="footer"/><address class="footer"><small>Generated on Thu Jan 20 2011 17:14:44 for Mgtaxa by&#160;
<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.2 </small></address>
</body>
</html>
