<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=iso-8859-1">
<title>Mgtaxa: MGT::Proj::CrisprApp::CrisprApp Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.4.6 -->
<div class="tabs">
  <ul>
    <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
    <li id="current"><a href="annotated.html"><span>Classes</span></a></li>
    <li><a href="pages.html"><span>Related&nbsp;Pages</span></a></li>
  </ul></div>
<div class="tabs">
  <ul>
    <li><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
    <li><a href="hierarchy.html"><span>Class&nbsp;Hierarchy</span></a></li>
    <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
  </ul></div>
<div class="nav">
<b>MGT</b>::<b>Proj</b>::<b>CrisprApp</b>::<a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html">CrisprApp</a></div>
<h1>MGT::Proj::CrisprApp::CrisprApp Class Reference</h1><!-- doxytag: class="MGT::Proj::CrisprApp::CrisprApp" --><a href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp-members.html">List of all members.</a><table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#51ea89ff680463872c8fc8ea8c81aa24">getDbSql</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#ceef3bc704046f0e488cdd7ce0dc897c">delDbSql</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#82338aeee20dc82aa166efeb3d35aa41">loadCrispr</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#b4673bfd9fa09943933181a00c7ca841">blastCr</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#81ad30589e546396ff13deb1a5b7db4e">blastCrOne</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#fe75d10fcdabd38012435ef17b37e85f">pilerCr</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#279e56dbf099b99efdf9a0fd1fd4428b">pilerCrOne</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#34ede177ed1a1fd7974463aac7d9fb29">blastToCoverage</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#c4297eebd63a5550312e133fab459034">loadMgtaxa</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#639bcb1470253f5146bc8e828fc34b1f">loadArrayBlastBin</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#dfbbb7fc7644c45aa3c3502940260401">parseArrayBlast</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#f2d174a1d3e38f8b3f2d47b65a5c9519">loadPilerArrSql</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#5efb2c9723dcf1b380e3d9a655a5bb14">loadCrisprSqlArr</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#8b95af2e19b80fa560f6ec1a894f9ed1">loadMic</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#a7b8ce2c664d2921bca1355de397b65a">loadMooreSql</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#ec2e7c463264d03dea9e08fef92a9f8d">loadClosedSql</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#d841b2c7a643a546801cdeecf4a3d6f1">mergeMicPathsSql</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#90e52a0ca79e28c4c02bcbc338f00883">makeMicNucFasta</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#9cfbf6f5ddf79177fe440b255e3ab442">loadCrArrayRangesSql</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#9c7aaa42a82573f41c0e2930129f4f57">loadCrArrayElemSql</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#9e377327e96450bee9dbf8b3aefec808">loadMicProtAnnotCrisprSql</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#e74465415c5d56ec22ca8ed06fa09ed5">annotMicCrisprGb</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#172fcd117894fed3cbc439659aae6a5e">loadMicProtHitsSql</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#d62dc25b2625ba187fae968a0a2fd46c">plotCrispr</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#e5ca96201bcae06434a59c0571c9ae94">plotCrisprOneBioGff</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#52abad8762653e5a7c8a709de4559c1d">plotCrisprOneGff</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classMGT_1_1Proj_1_1CrisprApp_1_1CrisprApp.html#2fade5136725d03893e7e5c520219b2f">plotCrisprOneBioGd</a></td></tr>

</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
<div class="fragment"><pre class="fragment">App-derived class for CRISPR array processing</pre></div> 
<p>
<hr><h2>Member Function Documentation</h2>
<a class="anchor" name="e74465415c5d56ec22ca8ed06fa09ed5"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::annotMicCrisprGb" ref="e74465415c5d56ec22ca8ed06fa09ed5" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::annotMicCrisprGb           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Add CRISPR arrays as annotation to Genbank format files</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="b4673bfd9fa09943933181a00c7ca841"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::blastCr" ref="b4673bfd9fa09943933181a00c7ca841" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::blastCr           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>kw</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">BLAST CRISPR arrays against different DBs in parallel</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="81ad30589e546396ff13deb1a5b7db4e"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::blastCrOne" ref="81ad30589e546396ff13deb1a5b7db4e" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::blastCrOne           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>kw</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">One job launched by blastCr()</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="34ede177ed1a1fd7974463aac7d9fb29"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::blastToCoverage" ref="34ede177ed1a1fd7974463aac7d9fb29" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::blastToCoverage           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Calculate how much spacers vs repeats each BLAST match hits.
@return record array with (id_arr,id_hit,coverage_spacers,coverage_repeats)</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="ceef3bc704046f0e488cdd7ce0dc897c"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::delDbSql" ref="ceef3bc704046f0e488cdd7ce0dc897c" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::delDbSql           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Call this to free a connection to SQL server if it will not be needed for extended period of time</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="51ea89ff680463872c8fc8ea8c81aa24"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::getDbSql" ref="51ea89ff680463872c8fc8ea8c81aa24" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::getDbSql           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Allocate (if necessary) and return a connection to SQL server</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="639bcb1470253f5146bc8e828fc34b1f"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::loadArrayBlastBin" ref="639bcb1470253f5146bc8e828fc34b1f" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::loadArrayBlastBin           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>inFile</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>minAlignLen</em> = <code>20</code>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>minBitScore</em> = <code>40</code>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>maxEvalue</em> = <code>1000</code>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>maxMism</em> = <code>1</code>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>out</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>debug</em> = <code>False</code></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Load into memory numpy record array previously created by parseArrayBlast().
The records will be filtered according to cutoff parameters.</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="ec2e7c463264d03dea9e08fef92a9f8d"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::loadClosedSql" ref="ec2e7c463264d03dea9e08fef92a9f8d" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::loadClosedSql           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Load meta-info about Genbank non-Moore closed marine genomes into SQL tables</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="9c7aaa42a82573f41c0e2930129f4f57"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::loadCrArrayElemSql" ref="9c7aaa42a82573f41c0e2930129f4f57" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::loadCrArrayElemSql           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>db</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Return numpy record array from SQL tables with every element of every CRISPR array.
id_seq is included, and records are sorted by (id_arr,begin).</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="9cfbf6f5ddf79177fe440b255e3ab442"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::loadCrArrayRangesSql" ref="9cfbf6f5ddf79177fe440b255e3ab442" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::loadCrArrayRangesSql           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>db</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Return numpy record array from SQL tables with total coord range of every CRISPR array</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="82338aeee20dc82aa166efeb3d35aa41"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::loadCrispr" ref="82338aeee20dc82aa166efeb3d35aa41" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::loadCrispr           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Import and filter original PILERCR output files and convert them into SQL tables</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="5efb2c9723dcf1b380e3d9a655a5bb14"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::loadCrisprSqlArr" ref="5efb2c9723dcf1b380e3d9a655a5bb14" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::loadCrisprSqlArr           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Load CRISPR from SQL tables into an array</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="c4297eebd63a5550312e133fab459034"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::loadMgtaxa" ref="c4297eebd63a5550312e133fab459034" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::loadMgtaxa           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Load MGTAXA assignments into SQL tables.
We parse the text output file, with each line like the one below:
1118658405050   rank=genus : name=\"\"\"Candidatus Pelagibacter\"\"\" : taxid=198251 &lt;&lt;-&gt;&gt;...</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="8b95af2e19b80fa560f6ec1a894f9ed1"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::loadMic" ref="8b95af2e19b80fa560f6ec1a894f9ed1" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::loadMic           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Load data from Mic genomes necessary to run our CRISPR pipeline on them</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="9e377327e96450bee9dbf8b3aefec808"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::loadMicProtAnnotCrisprSql" ref="9e377327e96450bee9dbf8b3aefec808" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::loadMicProtAnnotCrisprSql           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Create SQL table with the most definitive name for CAS genes and their coordinates</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="172fcd117894fed3cbc439659aae6a5e"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::loadMicProtHitsSql" ref="172fcd117894fed3cbc439659aae6a5e" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::loadMicProtHitsSql           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Load reliable protein HMM hits for Moore and other closed genomes into SQL tables</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="a7b8ce2c664d2921bca1355de397b65a"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::loadMooreSql" ref="a7b8ce2c664d2921bca1355de397b65a" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::loadMooreSql           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Load meta-info about Moore genomes into SQL tables</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="f2d174a1d3e38f8b3f2d47b65a5c9519"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::loadPilerArrSql" ref="f2d174a1d3e38f8b3f2d47b65a5c9519" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::loadPilerArrSql           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>db</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>idGenArr</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>inserterArr</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>idGenArrElem</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>inserterArrElem</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>arr</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Load array dict prepared by importPiler() into SQL tables.
</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="90e52a0ca79e28c4c02bcbc338f00883"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::makeMicNucFasta" ref="90e52a0ca79e28c4c02bcbc338f00883" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::makeMicNucFasta           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Export microbial GB nucl sequence into FASTA, store seq IDs into SQL table along the way, and split files for Piler run</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="d841b2c7a643a546801cdeecf4a3d6f1"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::mergeMicPathsSql" ref="d841b2c7a643a546801cdeecf4a3d6f1" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::mergeMicPathsSql           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Merge tables *_path with files paths for different sources of data, such as Moore and other Genbank marine genomes.
@post Results are in mic_path table</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="dfbbb7fc7644c45aa3c3502940260401"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::parseArrayBlast" ref="dfbbb7fc7644c45aa3c3502940260401" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::parseArrayBlast           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>inFile</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>minAlignLen</em> = <code>10</code>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>minBitScore</em> = <code>10</code>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>maxEvalue</em> = <code>100</code>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>maxMism</em> = <code>100</code>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>out</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>debug</em> = <code>False</code></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Parse BLAST XML output and return as numpy record array.
The records will be filtered according to cutoff parameters.</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="fe75d10fcdabd38012435ef17b37e85f"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::pilerCr" ref="fe75d10fcdabd38012435ef17b37e85f" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::pilerCr           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>kw</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Find CRISPR arrays with PilerCr within several partitions of the input sequence in parallel</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="279e56dbf099b99efdf9a0fd1fd4428b"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::pilerCrOne" ref="279e56dbf099b99efdf9a0fd1fd4428b" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::pilerCrOne           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>kw</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">One job launched by pilerCrOne()</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="d62dc25b2625ba187fae968a0a2fd46c"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::plotCrispr" ref="d62dc25b2625ba187fae968a0a2fd46c" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::plotCrispr           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname1" valign="top" nowrap> <em>self</em>          </td>
          <td class="md" valign="top">&nbsp;)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Use Bio.Graphics.GenomeDiagram to generate pictures of CRISPR genes and arrays from pre-processed Genbank files</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="2fade5136725d03893e7e5c520219b2f"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::plotCrisprOneBioGd" ref="2fade5136725d03893e7e5c520219b2f" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::plotCrisprOneBioGd           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>seqRec</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>outGraphFileRoot</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Use Bio.Graphics.GenomeDiagram to generate pictures of CRISPR genes and arrays from one SeqRecord from a pre-processed Genbank file</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="e5ca96201bcae06434a59c0571c9ae94"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::plotCrisprOneBioGff" ref="e5ca96201bcae06434a59c0571c9ae94" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::plotCrisprOneBioGff           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>seqRec</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>outGraphFileRoot</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Generate GFF files of CRISPR genes and arrays from one SeqRecord from a pre-processed Genbank file, using Bio sandbox module</pre></div>     </td>
  </tr>
</table>
<a class="anchor" name="52abad8762653e5a7c8a709de4559c1d"></a><!-- doxytag: member="MGT::Proj::CrisprApp::CrisprApp::plotCrisprOneGff" ref="52abad8762653e5a7c8a709de4559c1d" args="" --><p>
<table class="mdTable" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">def MGT::Proj::CrisprApp::CrisprApp::plotCrisprOneGff           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">&nbsp;</td>
          <td class="mdname" nowrap> <em>self</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>seqRec</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>&nbsp;</td>
          <td class="mdname" nowrap> <em>outGraphFileRoot</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
<div class="fragment"><pre class="fragment">Generate GFF files and graphics for CRISPR genes and arrays from one SeqRecord from a pre-processed Genbank file</pre></div>     </td>
  </tr>
</table>
<hr>The documentation for this class was generated from the following file:<ul>
<li>mgtaxa/MGT/Proj/CrisprApp.py</ul>
<hr size="1"><address style="align: right;"><small>Generated on Sat May 8 15:32:20 2010 for Mgtaxa by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.4.6 </small></address>
</body>
</html>
